File: backend/GameApp.Domain/ValueObjects/Characters/CharacterName.cs
namespace GameApp.Domain.ValueObjects;

// Value Object of Character Name for character's entities
public class CharacterName
{
    private static readonly string _messageIfEmpty = "Character name cannot be empty";
    private readonly string Name;



    // Constructor
    public CharacterName(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ArgumentException(_messageIfEmpty, nameof(value));

        Name = value;
    }

    public string GetName() => Name;


    public CharacterName SetName(string newName)
    {
        if (string.IsNullOrWhiteSpace(newName))
            throw new ArgumentException(_messageIfEmpty, nameof(newName));

        return new CharacterName(newName);
    }

    public override bool Equals(object? obj)
    {
        if (obj is not CharacterName other)
            return false;

        return Name == other.Name;
    }

    public override int GetHashCode() => Name.GetHashCode();

    public override string ToString() => Name;
}

File: frontend/gameapp-frontend/src/domain/value-objects/characters/character-name.js
// Value Object for Character Name
export class CharacterName {
    static ERROR_EMPTY = "Character name cannot be empty";

    constructor(name) {
        this.#validateName(name);
        this._name = name;
    }

    #validateName(name) {
        if (!name || name.trim().length === 0) {
            throw new Error(CharacterName.ERROR_EMPTY);
        }
    }

    // getter
    get name() {
        return this._name;
    }

    // setter
    setName(newName) {
        this.#validateName(newName);
        return new CharacterName(newName);
    }

    equals(other) {
        if (!(other instanceof CharacterName)) return false;
        return this._name === other._name;
    }

    toString() {
        return this._name;
    }
}